A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -0,0 +1,251 @@
+
+#### Java
+
+`OrcaOobUtil` is available in `io.grpc.xds.orca` but currently not in use. This needs to be moved to a new package and shared with `weighted_round_robin`, which implements `ForwardingLoadBalancerHelper` that takes the helper from `OrcaOobUtil`. The listener takes the reference to the subchannel to which it updates the weight to in `onLoadReport()`. A new API is needed to set and unset the listener in `OrcaOobUtil` to toggle the OOB load report processing.
+
+#### Go
+
+`orca.OOBListener` is available for OOB load reporting. `weighted_round_robin` extends it and attach it to the subchannel to which it updates the weight in `onLoadReport()`. A new API is needed to support enabling and disabling OOB load reporting.
+
+
+### xDS Integration
+
+`weighted_round_robin` is enabled as the endpoint picking policy of `WrrLocality` of [gRFC A52]. The implementation is available in C++ and Java but not in Go yet.
+
+`weighted_round_robin` is added as a new LB policy per [gRFC A52] and is added to `XdsLbPolicyRegistry`. The name is prefixed with `ClientSide` to avoid confusion with the existing weighted round robin support via `ROUND_ROBIN`.
+
+```textproto
+package envoy.extensions.load_balancing_policies.client_side_weighted_round_robin.v3;
+
+message ClientSideWeightedRoundRobin {
+   google.protobuf.Duration blackout_period = 1;
+   google.protobuf.Duration weight_expiration_period = 2;
+   bool enable_oob_load_report = 2;
+   google.protobuf.Duration oob_reporting_period = 3;
+   google.protobuf.Duration weight_update_period = 4;
+}
+```
+
+Example C++ code snippet that adds `weighted_round_robin` to `XdsLbPolicyRegistry`.
+```c
+//envoy.extensions.load_balancing_policies.client_side_weighted_round_robin.v3.ClientSideWeightedRoundRobin
+class ClientSideWeightedRounbRobinLbPolicyConfigFactory
+    : public XdsLbPolicyRegistry::ConfigFactory { … };
+
+XdsLbPolicyRegistry::XdsLbPolicyRegistry() {
+  // …
+  policy_config_factories_.emplace(
+      ClientSideWeightedRounbRobinLbPolicyConfigFactory::Type(),
+      absl::make_unique<ClientSideWeightedRounbRobinLbPolicyConfigFactory>());
+}
+```
+
+In the xDS configuration for the client, `client_side_weighted_round_robin` is wrapped in `wrr_locality` as a child policy. Below is an example xDS configuration with `client_side_weighted_round_robin`. The config includes `round_robin` as a backup policy to fallback in case the client does not support `client_side_weighted_round_robin`. On the client the new xDS LB hierarchy in [gRFC A52] creates the gRPC LB policy `weighted_round_robin` as the endpoint-picking policy.
+
+```json
+{ // Cluster
+  name: "..."
+  type: EDS
+  load_balancing_policy: {
+    policies: [
+      typed_extension_config: {
+        name: "..."
+        typed_config: {
+          "@type": "type.googleapis.com/envoy.extensions.load_balancing_policies.wrr_locality.v3.WrrLocality"
```

Comment:
Suggest writing this in the normal textproto notation for `Any` fields:

```
typed_config {
  [type.googleapis.com/envoy.extensions.load_balancing_policies.wrr_locality.v3.WrrLocality] {
    endpoint_picking_policy: {
      ...
    }
  }
}
```

Reply:
Done.